core/src/test/java/org/infinispan/partitionhandling/PartitionHigherTopologyIDTest.java

```diff
@@ -0,0 +1,87 @@
+ */
+@Test(groups = "functional", testName = "conflict.impl.MultipleCachesDuringConflictResolutionTest")
+public class PartitionHigherTopologyIDTest extends BasePartitionHandlingTest {
+
+    private final PartitionDescriptor p0;
+    private final PartitionDescriptor p1;
+
+    public PartitionHigherTopologyIDTest() {
+        this.p0 = new PartitionDescriptor(0, 1);
+        this.p1 = new PartitionDescriptor(2, 3);
+        this.partitionHandling = PartitionHandling.ALLOW_READ_WRITES;
+    }
+
+    public void testPartitionWithDifferentTopologyIds() {
+        TestingUtil.waitForNoRebalance(caches());
+
+        log.tracef("split test");
+        splitCluster(p0.getNodes(), p1.getNodes());
+        TestingUtil.waitForNoRebalance(caches());
+
+        ClusterTopologyManager manager = ComponentRegistry.of(cache(0)).getComponent(ClusterTopologyManager.class);
+
+        // We toggle the availability to increase the topology id
+        log.tracef("Forcing available status");
+        manager.forceAvailabilityMode("defaultcache", AvailabilityMode.AVAILABLE);
+
+        TestingUtil.waitForNoRebalance(getPartitionCaches(p0));
+
+        log.tracef("Forcing degraded status");
+        manager.forceAvailabilityMode("defaultcache", AvailabilityMode.DEGRADED_MODE);
+
+        TestingUtil.waitForNoRebalance(getPartitionCaches(p0));
+
+        log.tracef("performMerge");
+        partition(0).merge(partition(1));
+        TestingUtil.waitForNoRebalance(caches());
+
+        LocalizedCacheTopology topology = cache(0).getAdvancedCache().getDistributionManager().getCacheTopology();
+        assertEquals(4, topology.getMembers().size());
+    }
+
+    public void testAsymmetricPartition() throws InterruptedException {
+        // Disable discovery
+        for (int i = 0; i < numMembersInCluster; i++) {
+            JChannel channel = channel(cache(i));
+            disableDiscoveryProtocol(channel);
+
+            // Effectively disable MERGE3 protocol
+            MERGE3 merge3 = channel.getProtocolStack().findProtocol(MERGE3.class);
+            merge3.setMaxInterval(60000);
+            merge3.setMinInterval(60000);
+        }
+
+        // Note we only split the first partition. The second "partition" will still have all nodes but its topology
+        // is not increased until it finally merges back together.
+        // Manager 0 view -> {0, 1}
+        // Manager 1 view -> {0, 1}
+        // Manager 2 view -> {0, 1, 2, 3}
+        // Manager 3 view -> {0, 1, 2, 3}
+        asymmetricSplitCluster(p0.getNodes());
+
+        TestingUtil.waitForNoRebalance(getPartitionCaches(p0));
+
+        log.fatal("Rebalances complete, merging back together!");
+
+        partition(0).merge(partition(1));
+
+        TestingUtil.waitForNoRebalance(caches());
+
+        LocalizedCacheTopology localizedCacheTopology = TestingUtil.extractCacheTopology(cache(0));
+        assertEquals(4, localizedCacheTopology.getMembers().size());
```

Comment:
Should we verify all members have the same topology? To avoid if there is some special case in the coordinator.

Reply:
Sure